packages/compute/src/vm.js

```diff
@@ -809,6 +829,139 @@ VM.prototype.request = function(reqOpts, callback) {
   });
 };
 
+/**
+ * Poll `getMetadata` to check the VM's status. This runs a loop to ping
+ * the API on an interval.
+ *
+ * Note: This method is automatically called when a `waitFor()` call
+ * is made.
+ *
+ * @private
+ */
+VM.prototype.startPolling_ = function() {
+  var self = this;
+
+  if (!this.hasActiveWaiters) {
+    return;
+  }
+
+  this.getMetadata(function(err, metadata) {
+    var now = new Date();
+    for (var index in self.waiters) {
+      if (err) {
+        self.waiters[index].callback(err, null);
+        self.waiters.splice(index, 1);
+      } else if (metadata && metadata.status === self.waiters[index].status) {
+        self.waiters[index].callback(null, metadata);
+        self.waiters.splice(index, 1);
+      } else {
+        if ((now.getTime() / 1000) - self.waiters[index].startTime >=
+              self.waiters[index].timeout) {
+          self.waiters[index].callback({
+            name: 'WaitForTimeout'
+          },null);
+          self.waiters.splice(index, 1);
+        }
+      }
+    }
+    if (self.waiters.length <= 0) {
+      self.hasActiveWaiters = false;
+    } else {
+      setTimeout(self.startPolling_.bind(self), WAIT_FOR_POLLING_INTERVAL);
+    }
+  });
+};
+
+/**
+ * This function will callback when the VM is in the specified state.
+ *
+ * Will time out after the specified time (default: 300 seconds).
+ *
+ * @param {string} status - The status to wait for. This can be:
+ *  - "PROVISIONING"
+ *  - "STAGING"
+ *  - "RUNNING"
+ *  - "STOPPING"
+ *  - "SUSPENDING"
+ *  - "SUSPENDED"
+ *  - "TERMINATED"
+ * @param {object=} options - Configuration object.
+ * @param {number} options.timeout - The number of seconds to wait until
+ *     timing out.
+ *     Default: `300`
+ *     Maximum: `600`
+ * @param {function} callback - The callback function.
+ * @param {object} callback.metadata - The instance's metadata.
+ * @param {?error} callback.err - An error returned while waiting
+ *     for the status.
+ *
+ * @example
+ * vm.waitFor('RUNNING', function(err, metadata, apiResponse) {
+ *   if (!err) {
+ *     // The VM is running.
+ *   }
+ * });
+ *
+ * vm.waitFor('STOPPED', options, function(err, apiResponse) {
+ *   if (!err) {
+ *     // The VM is stopped.
+ *   }
+ * });
+ *
+ * //-
+ * // By default, `waitFor` will timeout after 300 seconds after waiting for
+ * // the desired state to occur. This can be changed to any number between
+ * // 0 and 600
+ * //-
+ * var options = {
+ *   timeout: 300
+ * };
+ *
+ * //-
+ * // If the callback is omitted, we'll return a Promise.
+ * //-
+ * vm.waitFor('RUNNING', options).then(function(data) {
+ *   var metadata = data[0];
+ * });
+ */
+VM.prototype.waitFor = function(status, options, callback) {
```

Comment:
We would still have to make a getMetadata call in order to check the status, and that happens immediately when we call startPolling_?